core/spam/vote_spam_policy.go

```diff
@@ -436,3 +436,24 @@ func (vsp *VoteSpamPolicy) PreBlockAccept(tx abci.Tx) (bool, error) {
 
 	return true, nil
 }
+
+func (vsp *VoteSpamPolicy) GetStats(partyID string) []Statistic {
+	vsp.lock.RLock()
+	defer vsp.lock.RUnlock()
+
+	partyStats := vsp.partyToVote[partyID]
+
+	stats := make([]Statistic, 0, len(partyStats))
+	bannedUntil := vsp.bannedParties[partyID]
+
+	for proposal, votes := range partyStats {
+		stats = append(stats, Statistic{
+			Name:        proposal,
+			Total:       votes,
+			Limit:       vsp.numVotes,
+			BannedUntil: bannedUntil,
```

Comment:
why is bannedUntil inside here? it should be global for the party

Reply:
It's because this is the object returned by anything implementing the spam.Policy interface. The spam.Engine will extract it out and provide only 1 in the response.

```
// The voting statistics by proposal for a given party for the current epoch
message VoteSpamStatistics {
  // List of statistics for proposals voted on by the party
  repeated VoteSpamStatistic statistics = 1;
  // If blocked the timestamp when the party will be unblocked as RFC3339Nano
  optional string banned_until = 2;
}

// The vote statistics for the voting spam policies
// which are calculated as a ratio of the total votes
// that have been rejected.
message VoteSpamStatistic {
  string proposal = 1;
  // The current vote count received from the party for the given proposal during this epoch
  uint64 count_for_epoch = 2;
  // The maximum number of votes per proposal allowed in an epoch
  uint64 max_for_epoch = 3;
}
```